ocamlPackages.memprof-limits: init at 0.2.1

Alizter opened the pull request:
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Alizter committed:
ocamlPackages.memprof-limits: init at 0.2.1

Signed-off-by: Ali Caglayan <alizter@gmail.com>

vbgl commented on pkgs/development/ocaml-modules/memprof-limits/default.nix:
```diff
@@ -0,0 +1,23 @@
+    homepage = "https://ocaml.org/p/memprof-limits/latest";
+    description =
+      "Memory limits, allocation limits, and thread cancellation for OCaml";
+    liscense = licenses.lgpl3;
```
```suggestion
    license = licenses.lgpl3;
```

Alizter committed:
Update pkgs/development/ocaml-modules/memprof-limits/default.nix

Co-authored-by: Vincent Laporte <vbgl@users.noreply.github.com>

Alizter commented:
Yes, it seems we need to restrict < 5.0.

Alizter committed:
ocamlPackages.memprof-limits: restrict to OCaml < 5.0

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter commented:
@vbgl I've added a check. Do you think it would be better if we introduced maximalOCamlVersion in buildDunePackage?

vbgl commented:
> Do you think it would be better if we introduced maximalOCamlVersion in buildDunePackage?

It sounds like a good idea. In particular, this would help avoiding silly mistakes such as swapping the arguments to `versionAtLeast`.

vbgl commented on pkgs/development/ocaml-modules/memprof-limits/default.nix:
```diff
@@ -0,0 +1,26 @@
+{ lib, buildDunePackage, fetchFromGitLab, ocaml }:
+
+if (lib.versionAtLeast "5.0.0" ocaml.version) then
```
```suggestion
if (lib.versionAtLeast ocaml.version "5.0.0") then
```

Alizter commented on pkgs/build-support/ocaml/dune.nix:
```diff
@@ -8,7 +8,8 @@ let Dune =
 
 if (args ? minimumOCamlVersion && lib.versionOlder ocaml.version args.minimumOCamlVersion) ||
-   (args ? minimalOCamlVersion && lib.versionOlder ocaml.version args.minimalOCamlVersion)
+   (args ? minimalOCamlVersion && lib.versionOlder ocaml.version args.minimalOCamlVersion) ||
```
Is it correct that our minimal ocaml version bound is strict?

vbgl commented:
Please leave the refactoring of `buildDunePackage` as a separate PR.

Alizter commented:
@vbgl OK, I'll put that in a separate PR then.

vbgl commented:
While you’re at it, please double-check what happens to `minimalOCamlVersion` and the like (i.e., arguments that are intended for dune wrapper and not for the underlying `mkDerivation`).

Alizter commented:
I've dropped the commit touching buildDunePackage (I'll handle that in another PR) and updated the version check. I am using versionOlder and the arguments should be in the correct order this time.

vbgl commented on pkgs/development/ocaml-modules/memprof-limits/default.nix:
```diff
@@ -0,0 +1,26 @@
+{ lib, buildDunePackage, fetchFromGitLab, ocaml }:
+
+if (lib.versionOlder ocaml.version "5.0.0") then
```
```suggestion
if !(lib.versionOlder ocaml.version "5.0.0") then
```

Alizter committed:
Update pkgs/development/ocaml-modules/memprof-limits/default.nix

Co-authored-by: Vincent Laporte <vbgl@users.noreply.github.com>
